<p>When a test fails due, for example, to infrastructure issues, you might want to ignore it temporarily.
But without some kind of notation about why the test is being disabled, it may never be reactivated.</p>

<p>This rule raises an issue for each disabled test that does not have a notation about why it was disabled.</p>

<h2>Noncompliant Code Example</h2>

<pre>
-- %test(my test)
-- %disabled
procedure test;
</pre>

<h2>Compliant Solution</h2>

<pre>
-- %test(my test)
procedure test;
</pre>

<h2>Exceptions</h2>

<p>The rule doesn't raise an issue if there is a comment in the <code>%disabled</code> annotation.</p>
